components/playlist/playlist_media_file_downloader.cc

```diff
@@ -0,0 +1,339 @@
+          source_media_files_dir.begin(), source_media_files_dir.end()));
+
+  const base::FilePath unified_media_file_path =
+      playlist_dir_path.Append(base::FilePath::StringType(
+          unified_media_file_name.begin(), unified_media_file_name.end()));
+
+  // Prepare empty target file to append.
+  base::DeleteFile(unified_media_file_path);
+  base::File unified_media_file(
+      unified_media_file_path,
+      base::File::FLAG_CREATE | base::File::FLAG_WRITE);
+  if (!unified_media_file.IsValid())
+    return false;
+  unified_media_file.Close();
+
+  const size_t kReadBufferSize = 1 << 16;  // 64KiB
+  char read_buffer[kReadBufferSize];
+  bool failed = false;
+  for (int i = 0; i < num_source_files; ++i) {
+    const base::FilePath media_file_source_path =
+        source_files_dir.Append(GetFileNameStringFromIndex(i));
+    if (!base::PathExists(media_file_source_path)) {
+      VLOG(2) << __func__
+              << " file not existed: " << media_file_source_path.value();
+      failed = true;
+      break;
+    }
+
+    if (!AppendToFileThenDeleteSource(media_file_source_path,
+                                      unified_media_file_path, read_buffer,
+                                      kReadBufferSize)) {
+      failed = true;
+      break;
+    }
+  }
+
+  // Delete source files dir.
+  base::DeletePathRecursively(source_files_dir);
+
+  if (failed) {
+    base::DeleteFile(unified_media_file_path);
+    return false;
+  }
+
+  return true;
+}
+
+}  // namespace
+
+PlaylistMediaFileDownloader::PlaylistMediaFileDownloader(
+    Delegate* delegate,
+    content::BrowserContext* context,
+    base::FilePath::StringType source_media_files_dir,
+    base::FilePath::StringType unified_media_file_name,
+    std::string media_file_path_key,
+    std::string create_params_path_key)
+    : delegate_(delegate),
+      url_loader_factory_(
+          context->content::BrowserContext::GetDefaultStoragePartition()
+              ->GetURLLoaderFactoryForBrowserProcess()),
+      request_helper_(std::make_unique<api_request_helper::APIRequestHelper>(
+          GetNetworkTrafficAnnotationTagForURLLoad(),
+          url_loader_factory_)),
+      source_media_files_dir_(source_media_files_dir),
+      unified_media_file_name_(unified_media_file_name),
+      media_file_path_key_(media_file_path_key),
+      create_params_path_key_(create_params_path_key),
+      weak_factory_(this) {}
+
+PlaylistMediaFileDownloader::~PlaylistMediaFileDownloader() = default;
+
+void PlaylistMediaFileDownloader::NotifyFail(const std::string& id) {
+  ResetDownloadStatus();
+  delegate_->OnMediaFileGenerationFailed(id);
+}
+
+void PlaylistMediaFileDownloader::NotifySucceed(
+    const std::string& id,
+    const std::string& media_file_path_key,
+    const std::string& media_file_path) {
+  ResetDownloadStatus();
+  delegate_->OnMediaFileReady(id, media_file_path_key, media_file_path);
+}
+
+void PlaylistMediaFileDownloader::GenerateSingleMediaFile(
```

Comment:
It seems videos and audios in a playlist are merged into a single file. if so, this will make it hard:
* to edit playlist, e.g. remove an item from the list.
* share same file across playlists

What can be benefit over these limits?

Reply:
Single file per each playlist item not whole playlist.
IIRC, youtube gives multiple sources for one clip. So, this method was introduced.

Reply:
Aha, yes. youtubedown.js seems to suggest multiple source based on resolutions and others. then, Maybe we can remove this as our script gives us one url.